model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/CacheDecorators.java

```diff
@@ -40,17 +42,31 @@ public static <K, V> AdvancedCache<K, V> localCache(Cache<K, V> cache) {
      * @param cache
      * @return Cache with the flags applied.
      */
-    public static <K, V> AdvancedCache<K, V> skipCacheLoaders(Cache<K, V> cache) {
-        return cache.getAdvancedCache().withFlags(Flag.SKIP_CACHE_LOAD, Flag.SKIP_CACHE_STORE);
+    public static <K, V> AdvancedCache<K, V> skipCacheLoadersIfRemoteStoreIsEnabled(Cache<K, V> cache) {
+        if (!getRemoteStores(cache).isEmpty()) {
+            // Disabling of the cache load and cache store is only needed when a remote store is used and handled separately.
+            return cache.getAdvancedCache().withFlags(Flag.SKIP_CACHE_LOAD, Flag.SKIP_CACHE_STORE);
+        } else {
+            // If there is no remote store, use write through/read through for all stores of the cache.
+            // Mixing remote and non-remote caches is not supported.
+            return cache.getAdvancedCache();
```

Comment:
Hi @ahus1 ,

Thanks for starting this PR, even if as you said it is not a recommended setup, several projects/people use it.

I understand the effect of removing `SKIP_CACHE_STORE` here, but I don't get all impacts of removing the `SKIP_CACHE_LOAD` too if no remote store is used. 

I already noticed performance issues while removing `SKIP_CACHE_LOAD` here. With hundred of thousands of user/client sessions, user API will become very slow : more than 30 seconds or timeout when searching a user in the admin console, several seconds to delete one by API ... 

It was making too many SQL queries to ISPN sessions/clientSessions tables, always fetching from database : 
`SELECT datum, id FROM "ispn_sessions" WHERE (timestamp > $1 OR timestamp < 0) AND segment IN ($2) `

It is just a hint on my past observations on that, I know that it is working fine without performance issues with this code and around ~ 2 millions user sessions : 

```suggestion
            return cache.getAdvancedCache(Flag.SKIP_CACHE_LOAD);
```

Regarding your comment about `preload` in ISPN XML config file, I also can confirm that it works fine without any performance or functional issue with the default `preload=false`, combined with an : 
`<state-transfer enabled="true" await-initial-transfer="false" />`

In a 3-node cluster with 2 owners for distributed-cache, it makes the node available quite quickly (a few seconds) and prevent timeout issues during initial state transfer which `240000`  default value is too short for a huge number of sessions, even when increasing the state-transfer `chunk-size`. Of course state-transfer timeout can also be increased in case of sync startup.

Reply:
How does `Flag.SKIP_CACHE_LOAD` effects this when there is no remote store to skip?

Reply:
Thank you for the update, I accepted the change. 

I agree that regular user handling will work (login, logout, refresh), but I doubt that listing all the sessions of for example a user will work as expected as from looking at the code I assume all sessions need to be in the internal Infinispan store. Therefore the preloading and state transfer enabled. Deferring the initial transfer will deliver wrong results until the transfer is complete, but that might be a trade off for some deployments.

Let's go with this version as an improvement, and possibly have future PR to enhance it even more.

Reply:
@martin-kanis - the "skip load" will prevent the "string-keyed-jdbc-store" to be searched when listing, therefore I think it is a good idea and actually makes this change smaller.

Reply:
Then I find name of the method confusing. It skips cache load every time, not just if remote store is enabled.

Reply:
yes exactly @martin-kanis that's why I was telling that for me the role of `skipCacheLoad` was really confusing, contrary to `skipCacheStore`  which is clearer.

You are right about consistency @ahus1 on startup, until user sessions have been loaded in internal cache. For example in admin console they won't be visible immediatly. 
But for some customers it can be an acceptable trade-off as you said.